README.md
Planning Biblio, Plugins Planning Hebdo
==============


Changelog.txt
***	Version 1.4.6	(Jérôme Combes, 17 avril 2015)		***
- Corrections :
Corrige les vulnérabilités Reflected XSS
Corrige les vulnérabilités Client Potential XSS

***	Version 1.4.5	(Jérôme Combes, 10 avril 2015)		***
- Améliorations :
Optimise la recherche des emplois du temps sur la page du planning et dans le menu permettant de placer les agents (gain de temps)

***	Version 1.4.3	(Jérôme Combes, 28 mars 2015)		***
- Améliorations :
Créé les dataTables avec la classe CJDataTable

***	Version 1.4.1	(Jérôme Combes, 17 mars 2015)		***
- Améliorations :
Initialisation du datatable de la page index.php dans fichier JS

***	Version 1.3	(Jérôme Combes, 17 janvier 2014)		***
- Améliorations :
Utilisation de DataTable, de DatePicker et du style JQuery-UI highlight et error pour les notifications
Dates au format JJ/MM/AAAA

***	Version 1.2.13	(Jérôme Combes, 9 décembre 2013)		***
- Correction de bugs :
Corrige le tri par dates dans le tableau administrateur des plannings de présence

***	Version 1.2.12	(Jérôme Combes, 26 novembre 2013)		***
- Correction de bugs :
N'affiche plus la possibilité de modifier des plannings de présence validés lorsque l'on utilise les 
périodes définies

***	Version 1.2.11	(Jérôme Combes, 14 novembre 2013)		***
- Améliorations :
Modification des paramétres de datePicker suite à l'utilisation de la traduction FR dans le planning

***	Version 1.2.10	(Jérôme Combes, 22 Octobre 2013)		***
- Améliorations :
Possibilité de trier le tableau des plannings de présence coté administrateur

***	Version 1.2.9	(Jérôme Combes, 11 Octobre 2013)		***
- Améliorations :
Modification des plannings de présence en cours d'année, même s'ils sont validés (copies)

***	Version 1.2.8	(Jérôme Combes, 27 Septembre 2013)		***
- Améliorations :
Supprime les plannings de présences des agents lors de leur suppression définitive

***	Version 1.2.7	(Jérôme Combes, 24 Septembre 2013)		***
- Améliorations :
Modification du titre "Récupération des samedis" -> "Récupération" dans le menu "Mon compte"

***	Version 1.2.6	(Jérôme Combes, 17 Septembre 2013)		***
- Améliorations :
Tri des plannings de présence par date de début,fin et nom de l'agent

***	Version 1.2.5	(Jérôme Combes, 17 Septembre 2013)		***
- Améliorations :
Envoie des notifications aux administrateur, y compris celui qui apporte les modifications, 
lors de la modification ou la validation d'un planning de présence

***	Version 1.2.4	(Jérôme Combes, 17 Septembre 2013)		***
- Améliorations :
Ajoute la possibilité de supprimer les plannings de présence

***	Version 1.2.2	(Jérôme Combes, 5 Septembre 2013)		***
- Améliorations :
Affichage des crédits de congés et récupérations dans le menun (si le plugin congés est installé)

***	Version 1.2.2	(Jérôme Combes, Août 2013)		***
- Infos
Nécessite la version 1.5.2 (ou plus) de planningBiblio
- Améliorations :
Affichage du nombres d'heures lors de la saisie du planning de présence.


modif.php
<?php
/*
Planning Biblio, Plugin planningHebdo Version 1.4.6
Licence GNU/GPL (version 2 et au dela)
Voir les fichiers README.md et LICENSE
Copyright (C) 2013-2015 - Jérôme Combes

Fichier : plugins/planningHebdo/modif.php
Création : 23 juillet 2013
Dernière modification : 22 avril 2015
Auteur : Jérôme Combes, jerome@planningbilbio.fr

Description :
Affiche la liste des plannings de présence pour l'administrateur
Page accessible à partir du menu administration/planning de présence
*/

require_once "class.planningHebdo.php";

// Recherche de la config
$p=new planningHebdo();
$p->getConfig();
$configHebdo=$p->config;

// Initialisation des variables
$id=filter_input(INPUT_GET,"id",FILTER_SANITIZE_NUMBER_INT);
$retour=filter_input(INPUT_GET,"retour",FILTER_SANITIZE_STRING);

$p=new planningHebdo();
$p->id=$id;
$p->fetch();
$debut1=$p->elements[0]['debut'];
$fin1=$p->elements[0]['fin'];
$debut1Fr=dateFr($debut1);
$fin1Fr=dateFr($fin1);

$perso_id=$p->elements[0]['perso_id'];
$temps=$p->elements[0]['temps'];
$valide=$p->elements[0]['valide'];
$remplace=$p->elements[0]['remplace'];

// Sécurité
$admin=in_array(24,$droits)?true:false;
if(!$admin and $perso_id!=$_SESSION['login_id']){
  echo "<div id='acces_refuse'>Accès refusé</div>\n";
  include "include/footer.php";
  exit;
}

// Modif autorisée si n'est pas validé ou si validé avec des périodes non définies (BSB).
// Dans le 2eme cas copie du planning avec modification des dates
$action="modif";
$modifAutorisee=true;
if(!$admin and $valide and $configHebdo['periodesDefinies']){
  $modifAutorisee=false;
}
if(!$admin and $valide){
  $action="copie";
}

// Informations sur l'agents
$p=new personnel();
$p->fetchById($perso_id);
$sites=$p->elements[0]['sites'];
?>

<!-- Formulaire Planning-->
<h3>Planning de présence</h3>
<?php echo "Planning de <b>".nom($perso_id,"prenom nom")."</b> du $debut1Fr au $fin1Fr";?>
<div id='planning'>
<?php
if(!$configHebdo['periodesDefinies']){
  echo "<form name='form1' method='post' action='index.php' onsubmit='return plHebdoVerifForm();'>\n";
}else{
  echo "<form name='form1' method='post' action='index.php' onsubmit='return verif_form(\"debut=date1;fin=date2Obligatoire\",\"form1\");'>\n";
}
?>
<input type='hidden' name='page' value='plugins/planningHebdo/valid.php' />
<input type='hidden' name='action' value='<?php echo $action; ?>' />
<input type='hidden' name='validation' value='0' />
<input type='hidden' name='retour' value='<?php echo $retour; ?>' />
<input type='hidden' name='id' value='<?php echo $id; ?>' />
<input type='hidden' name='perso_id' value='<?php echo $perso_id; ?>' />
<input type='hidden' name='valide' value='<?php echo $valide; ?>' />
<input type='hidden' name='remplace' value='<?php echo $remplace; ?>' />

<!-- Affichage des tableaux avec la sélection des horaires -->
<?php
switch($config['nb_semaine']){
  case 2	: $cellule=array("Semaine Impaire","Semaine Paire");		break;
  case 3	: $cellule=array("Semaine 1","Semaine 2","Semaine 3");		break;
  default 	: $cellule=array("Jour");					break;
}
$fin=$config['Dimanche']?array(8,15,22):array(7,14,21);
$debut=array(1,8,15);
$jours=Array("Lundi","Mardi","Mercredi","Jeudi","Vendredi","Samedi","Dimanche");
?>

<?php
for($j=0;$j<$config['nb_semaine'];$j++){
  echo "<br/>\n";
  echo "<table border='1' cellspacing='0' id='tableau{$j}' >\n";
  echo "<tr style='text-align:center;'><td style='width:150px;'>{$cellule[$j]}</td><td style='width:150px;'>Heure d'arrivée</td>";
  echo "<td style='width:150px;'>Début de pause</td><td style='width:150px;'>Fin de pause</td>";
  echo "<td style='width:150px;'>Heure de départ</td>";
  if($config['Multisites-nombre']>1){
    echo "<td>Site</td>";
  }
  echo "<td style='width:150px;'>Temps</td>";
  echo "</tr>\n";
  for($i=$debut[$j];$i<$fin[$j];$i++){
    $k=$i-($j*7)-1;
    echo "<tr style='text-align:center;'><td>{$jours[$k]}</td>";
    if($modifAutorisee){
      echo "<td>".selectTemps($i-1,0,null,"select")."</td><td>".selectTemps($i-1,1,null,"select")."</td>";
      echo "<td>".selectTemps($i-1,2,null,"select")."</td><td>".selectTemps($i-1,3,null,"select")."</td>";
    }
    else{
      echo "<td id='temps_".($i-1)."_0'>".heure2($temps[$i-1][0])."</td><td id='temps_".($i-1)."_1'>".heure2($temps[$i-1][1])."</td>";
      echo "<td id='temps_".($i-1)."_2'>".heure2($temps[$i-1][2])."</td><td id='temps_".($i-1)."_3'>".heure2($temps[$i-1][3])."</td>";
    }
    if($config['Multisites-nombre']>1){
      echo "<td><select name='temps[".($i-1)."][4]'>\n";
      if(count($sites)>1){
	echo "<option value=''>&nbsp;</option>\n";
      }
      foreach($sites as $site){
	$selected=$temps[$i-1][4]==$site?"selected='selected'":null;
	echo "<option value='$site' $selected >{$config["Multisites-site{$site}"]}</option>\n";
      }
      echo "</select></td>";
    }
    echo "<td id='heures_{$j}_$i'></td>\n";
    echo "</tr>\n";
  }
  echo "</table>\n";
  echo "Nombre d'heures : <font id='heures_{$j}' style='font-weight:bold;'>&nbsp;</font><br/>\n";
}

if(!$modifAutorisee){
  echo "<p><b class='important'>Vos horaires ont été validés.</b><br/>Pour les modifier, contactez votre chef de service.</p>\n";
}
elseif($valide and !$admin){
  echo "<p><b class='important'>Vos horaires ont été validés.</b><br/>Si vous souhaitez les changer, modifiez la date de début et/ou de fin d'effet.<br/>";
  echo "Vos nouveaux horaires seront enregistrés et devront être validés par un administrateur.<br/>";
  echo "Les anciens horaires seront conservés en attendant la validation des nouveaux.</p>\n";
}
elseif($valide and $admin and !$configHebdo['periodesDefinies']){
  echo "<p style='width:850px;text-align:justify;'><b class='important'>Ces horaires ont été validés.</b><br/>";
  echo "En tant qu'administrateur, vous pouvez les modifier et les enregistrer en tant que copie.<br/>";
  echo "Dans ce cas, modifiez la date de début et/ou de fin d'effet. ";
  echo "Les nouveaux horaires seront enregistrés et devront ensuite être validés. ";
  echo "Les anciens horaires seront conservés en attendant la validation des nouveaux.<br/>";
  echo "Vous pouvez également les enregistrer directement mais dans ce cas, vous ne conserverez pas les anciens horaires.</p>\n";
}
elseif($valide and $admin and $configHebdo['periodesDefinies']){
  echo "<p style='width:850px;text-align:justify;'><b class='important'>Ces horaires ont été validés.</b><br/>";
  echo "En tant qu'administrateur, vous avez toujours la possibilité de les modifier et de les valider.</p>\n";
}

// Choix de la période d'utilisation et validation
if($modifAutorisee and !$configHebdo['periodesDefinies']){
  echo "<br/><b>Choisissez la période d'utilisation et validez</b> :\n";
}

echo "<table style='width:900px;'>\n";
if(!$configHebdo['periodesDefinies']){
  echo <<<EOD
    <tr>
    <td>Date de début</td>
    <td><input type='text' name='debut' value='$debut1Fr' class='datepicker'/></td>
    <td>Date de fin</td>
    <td><input type='text' name='fin' value='$fin1Fr' class='datepicker'/></tr>
EOD;
}
else{
  echo "<tr><td><input type='hidden' name='debut' value='$debut1'/></td>\n";
  echo "<td><input type='hidden' name='fin' value='$fin1'/></td></tr>\n";
}
echo "<tr><td colspan='4' style='padding-top:20px;'>\n";
echo "<input type='button' value='Retour' onclick='location.href=\"index.php?page=plugins/planningHebdo/$retour\";' class='ui-button' />\n";

if($admin){
  echo "<input type='submit' value='Enregistrer les modifications SANS valider' style='margin-left:30px;' class='ui-button' />\n";
  if(!$configHebdo['periodesDefinies']){
    echo "<input type='button' value='Enregistrer et VALIDER'  style='margin-left:30px;' onclick='document.forms[\"form1\"].validation.value=1;if(plHebdoVerifForm()){document.forms[\"form1\"].submit();}' class='ui-button' />";
  }else{
    echo "<input type='button' value='Enregistrer et VALIDER'  style='margin-left:30px;' onclick='document.forms[\"form1\"].validation.value=1;document.forms[\"form1\"].submit();' class='ui-button' />";
  }
  if($valide and !$configHebdo['periodesDefinies']){
    echo "<input type='button' value='Enregistrer une copie' style='margin-left:30px;' onclick='$(\"input[name=action]\").val(\"copie\");$(\"form[name=form1]\").submit();' class='ui-button' />\n";
  }
  echo "</td></tr>\n";
}
elseif($modifAutorisee){
  echo "<input type='submit' value='Enregistrer les modifications' style='margin-left:30px;' class='ui-button' />\n";
}

?>
</table>

</form>
<script type='text/JavaScript'>
$("document").ready(function(){plHebdoCalculHeures2();});
$(".select").change(function(){plHebdoCalculHeures($(this),"");});
</script>

</div> <!-- Planning -->